packages/graphback-datasync/src/providers/DatasyncMongoDBDataProvider.ts

```diff
@@ -22,45 +23,156 @@ export class DataSyncMongoDBDataProvider<Type = any> extends MongoDBDataProvider
       throw e;
     });
     this.coerceTSFields = true;
+    this.conflictEngine = conflictEngine;
   }
 
   public async create(data: any, context: GraphbackContext): Promise<Type> {
     data._deleted = false;
+    if (this.conflictEngine !== undefined) {
+      const { next, conflictFieldName } = this.conflictEngine;
+      data[conflictFieldName] = next(undefined, data);
+    }
 
     return super.create(data, context);
   }
 
   public async update(data: any, context: GraphbackContext): Promise<Type> {
-    const conflict = await this.checkForConflicts(data, context);
 
-    if (conflict !== undefined) {
-      throw new ConflictError(conflict);
+    const { idField } = getDatabaseArguments(this.tableMap, data);
+
+    if (!idField.value) {
+      throw new NoDataError(`Cannot update ${this.collectionName} - missing ID field`)
     }
 
-    // TODO use findOneAndUpdate to check consistency afterwards
-    return super.update(data, context);
+    const objectId = new ObjectId(idField.value);
+
+    const startTime = Date.now();
+
+    const RETRY_LIMIT = 3000; // 3 seconds
+
+    while ((Date.now() - startTime) < RETRY_LIMIT) {   // Commented for debugging
+    // while (Date.now()) {
+
+      this.fieldTransformMap[TransformType.UPDATE]
+        .forEach((f: FieldTransform) => {
+          data[f.fieldName] = f.transform(data[f.fieldName]);
+        });
+
+      const updateFilter: any = { _id: objectId, _deleted: { $ne: true } };
+      const updateSets = Object.assign({}, data);
+      // eslint-disable-next-line @typescript-eslint/tslint/config
+      delete updateSets[this.tableMap.idField];
+
+      if (this.conflictEngine !== undefined) {
+        
+        const { conflictFieldName, next } = this.conflictEngine;
+        const currentState = updateSets[conflictFieldName];
+        updateFilter[conflictFieldName] = currentState;
+
+        // eslint-disable-next-line @typescript-eslint/tslint/config
+        delete updateSets[conflictFieldName]
+
+        updateSets[conflictFieldName] = next(currentState, updateSets);
```

Comment:
nextState should have fieldName information inside IMHO.
Entire implementation is leaking some concepts.

I would recommend to:

1) Not introduce state providers yet (as this requires understanding off how the state will change in real scenarios)

2) Adapt existing implementation (obviously there are some issue there so we can chat about this)
https://offix.dev/docs/next/conflict-server